tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+    assert_script_run("ssh root\@$vm './${driver_file} -s'");
+    assert_script_run("ssh root\@$vm 'nvidia-smi'");
+    record_info("vGPU Grid driver is installed successfully in $vm");
+    # Verify
+    assert_script_run("ssh root\@$vm 'lsmod | grep -i nvidia'");
+}
+
+sub detach_vgpu_from_guest {
+    my ($slot, $vm) = @_;
+    assert_script_run("ssh root\@$vm 'poweroff'");
+    script_retry("virsh domstate $vm | grep 'shut off'", timeout => 60, delay => 5, retry => 3, die => 0);
+    assert_script_run("virt-xml $vm --remove-device --hostdev type=mdev,address.slot=$slot");
+    assert_script_run("virsh start $vm");
+    record_info("vGPU has been removed successfully from $vm");
+    wait_guest_online($vm);
+}
+
+sub remove_vgpu {
+    my $uuid = shift;
+    assert_script_run("cd `find /sys/devices -type d -name $uuid`");
+    assert_script_run("echo '1' > remove");
+    script_run('cd ../mdev_supported_types/; for i in *; do echo "$i" $(cat $i/name) available: $(cat $i/avail*); done');
+    die "Fatal: vGPU device $uuid is still alive!" if script_run("lsmdev | grep $uuid") == 0;
+    record_info("vGPU removed", "$uuid has been removed from the host");
+}
+
+sub save_vgpu_device_status_logs {
+    my $vm = shift;
+
+    #vm configuration file
+    enter_cmd("virsh dumpxml $vm > $log_dir/${vm}.xml");
+
+    my $log_file = "log.txt";
+    #logging device information in guest
+    if (script_run("virsh domstate $vm | grep running") == 0) {
+        script_run "echo `date` > $log_file";
+        script_run "echo '***** Status & logs inside $vm *****' >> $log_file";
+        my $debug_script = "vgpu_guest_logging.sh";
+        download_script($debug_script, machine => $vm, proceed_on_failure => 1);
+        script_run("timeout 25 ssh root\@$vm \"~/$debug_script\" >> $log_file 2>&1");
+        # Save nvidia driver installer logs
+        script_run("rsync root\@$vm:/var/log/nvidia-installer.log ${log_dir}/${vm}_nvidia-installer.log");
+        script_run "mv $log_file $log_dir/${vm}_vgpu_device_status.txt";
+    }
+}
+
+sub post_fail_hook {
+    my $self = shift;
+    diag("Module vgpu post fail hook starts.");
+
+    # Make sure sol console is ok
+    enter_cmd "echo DONE > /dev/$serialdev";
+    reconnect_when_ssh_console_broken unless defined(wait_serial('DONE', timeout => 30));
+
+    my $log_file = "$log_dir/host_vgpu_device_status.txt";
+    print_cmd_output_to_file("virsh list --all", $log_file);
+    print_cmd_output_to_file("nvidia-smi", $log_file);
+    my $cmd = "journalctl --cursor-file /tmp/cursor.txt | grep -r -e 'kernel:' -e gpu -e nvidia | grep -v bash_history";
+    script_run("echo '' >> $log_file");
+    script_run("$cmd >> $log_file");
+    script_run("cp /var/log/nvidia-installer.log $log_dir");
+    save_vgpu_device_status_logs($_) foreach (keys %virt_autotest::common::guests);
+    upload_virt_logs($log_dir, "vgpu");
+    $self->SUPER::post_fail_hook;
+    restore_original_guests() unless get_var('VGPU_SHORT_TEST');
```

Comment:
When test fails and there is problem to debug, with this post_fail_hook, will we have the failure scene to check? Seems environment is broken.

Reply:
Do you mean that `restore_original_guests()` flush the environment at the point of failures? Yes, it does. Apart from other feature tests, sriov and vgpu tests set `test_flags => 0` which will not stop tests even if there are failures, so restoring guests are necessary. That also means collecting and uploading enough logs during failure are important.

Reply:
Ok, leave it to you to decide :) , since you know better than me if the logs are enough. If it happens that we need to rerun to reproduce environment when test fails, it is also in time to change then.